File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        // left pointer
        int l = 0;
        
        String res = "";

        // hashmap stores unique char and counts of t
        Map<Character, Integer> map = new HashMap<>();
        for (int i = 0; i < t.length(); i++)
            map.put(t.charAt(i), map.getOrDefault(t.charAt(i), 0) + 1);
        
        // unique char count of t
        int tCount = map.size();
        
        // hashmap for the window
        Map<Character, Integer> window = new HashMap<>();        
        
        // unique char count of window
        int windowCount = 0;
        
        for (int r = 0; r < s.length(); r++) {
            if (!map.containsKey(s.charAt(r)))
                continue;
            
            // update window hashmap
            window.put(s.charAt(r), window.getOrDefault(s.charAt(r), 0) + 1);
            
            // update windowCount if the count of char at r matches
            if (window.get(s.charAt(r)).intValue() == map.get(s.charAt(r)).intValue())
                windowCount++;
            
            // move l++ if the window contians all chars in t
            while (l <= r && windowCount == tCount) {
                if (res.equals(""))
                    res = s.substring(l, r + 1);
                else {
                    if (s.substring(l, r + 1).length() < res.length())
                        res = s.substring(l, r + 1);
                }
                window.put(s.charAt(l), window.getOrDefault(s.charAt(l), 0) - 1);
                if (map.containsKey(s.charAt(l)) && window.get(s.charAt(l)).intValue() < map.get(s.charAt(l)).intValue())
                    windowCount--;
                l++;
            }
        }
        
        return res;
    }
}

File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        res = ""
        if len(s) < len(t):
            return res
        
        # build a hashmap for the chars in t 
        hmap = {}
        for i in range(len(t)):
            if t[i] not in hmap:
                hmap[t[i]] = 1
            else:
                hmap[t[i]] += 1
        
        # window pointers and hashmap for the chars in the window && in t
        l = 0
        r = 0
        windowMap = {}
        windowCount = 0
        mapCount = len(hmap)

        while r < len(s):
            if s[r] not in hmap:
                r += 1
            else:
                windowMap[s[r]] = windowMap.get(s[r], 0) + 1
                if windowMap[s[r]] == hmap[s[r]]:
                    windowCount += 1

                while l <= r and windowCount == mapCount:
                    if res == "" or len(res) > (r - l + 1):
                        res = s[l:r+1]
                    if s[l] in hmap:
                        windowMap[s[l]] -= 1
                    if s[l] in hmap and windowMap[s[l]] < hmap[s[l]]:
                        windowCount -= 1
                    l += 1
                r += 1
            
        return res